src/Tidy/Codegen/Monad.purs

```diff
@@ -91,11 +104,35 @@ data CodegenImport
 derive instance Eq CodegenImport
 derive instance Ord CodegenImport
 
+-- | `import Foo` = `OpenHiding Set.empty`
+-- | `import Foo hiding (bar)` = `OpenHiding (Set.singleton bar)`
+-- | `import Foo (bar)` = `ClosedImporting (NES.singleton bar)`
+-- |
+-- | If a module is imported multiple times unqualified,
+-- | import forms on the left will be overridden by import forms on the right
+-- | ```
+-- | ClosedImport _ < OpenImport Set.empty < OpenImport nonEmptySet
+-- | ```
+data UnqualifiedImportModule
+  = ClosedImporting (NonEmptySet CodegenImport)
+  | OpenHiding (Set CodegenImport)
+
+derive instance eqUnqualifiedImportModule :: Eq UnqualifiedImportModule
+derive instance ordUnqualifiedImportModule :: Ord UnqualifiedImportModule
+
 type CodegenState e =
   { exports :: Set CodegenExport
-  , importsOpen :: Set ModuleName
-  , importsFrom :: Map ModuleName (Set (CodegenImport))
-  , importsQualified :: Map ModuleName (Set ModuleName)
+  , importsUnqualified :: Map ModuleName UnqualifiedImportModule
+  -- | `import Foo as Bar` = `Map.singleton "Foo" $ Map.singleton "Bar" Nothing`
+  -- | `import Foo (baz) as Bar` = `Map.singleton "Foo" $ Map.singleton "Bar" $ Just $ NES.singleton "baz"`
+  -- |
+  -- | If a module is imported multiple times qualified,
+  -- | import forms on the left will be overridden by import forms on the right
+  -- | ```
+  -- | let map1 = Map.singleton
+  -- | (map1 "foo" (map1 "bar" $ Set.singleton "baz")) < (map1 "foo" (map1 "bar" Set.empty))
+  -- | ```
+  , importsQualified :: Map ModuleName (Map ModuleName (Set CodegenImport))
```

Comment:
This is formatted as doc comments, but AFAIK the doc generator will ignore this? Is this comment for lib maintainers? If it's for maintainers I guess I would just omit this, and instead document it with the algorithm if it's particularly subtle (I'm not sure it is).